tempesta_fw/http.c

```diff
@@ -2641,128 +2780,339 @@ tfw_http_hm_drop_resp(TfwHttpResp *resp)
+		r = tfw_http_parse_skb(conn, &skb, &off, &data_up);
 		switch (r) {
 		default:
-			TFW_ERR("Unrecognized HTTP request "
-				"parser return code, %d\n", r);
+			TFW_ERR("Unrecognized HTTP request parser return code "
+				"%d\n", r);
 			BUG();
 		case TFW_BLOCK:
 			TFW_DBG2("Block invalid HTTP request\n");
 			TFW_INC_STAT_BH(clnt.msgs_parserr);
-			tfw_client_drop(req, 400, "failed to parse request");
-			return TFW_BLOCK;
+			r = tfw_client_drop(req, 400, "failed to parse request");
+			goto err;
+
+		case TFW_SPLIT:
 		case TFW_POSTPONE:
 			r = tfw_gfsm_move(&conn->state, TFW_HTTP_FSM_REQ_CHUNK,
 					  &data_up);
 			TFW_DBG3("TFW_HTTP_FSM_REQ_CHUNK return code %d\n", r);
-			if (r == TFW_BLOCK) {
-				TFW_INC_STAT_BH(clnt.msgs_filtout);
-				tfw_client_block(req, 403, "postponed"
-					       " request has been"
-					       " filtered out");
-				return TFW_BLOCK;
+			if (likely(r != TFW_BLOCK)) {
+				r = TFW_POSTPONE;
+				break;
 			}
-			/*
-			 * TFW_POSTPONE status means that parsing succeeded
-			 * but more data is needed to complete it. Lower layers
-			 * just supply data for parsing. They only want to know
-			 * if processing of a message should continue or not.
-			 */
-			return TFW_PASS;
+
+			TFW_INC_STAT_BH(clnt.msgs_filtout);
+			r = tfw_client_block(
+				req, 403,
+				"postponed request has been filtered out");
+			goto err;
+
 		case TFW_PASS:
-			/*
-			 * The request is fully parsed,
-			 * fall through and process it.
-			 */
 			BUG_ON(!(req->flags & TFW_HTTP_F_CHUNKED)
 			       && (req->content_length != req->body.len));
+			break;
 		}
+	}
+	stream_mode = tfw_http_msg_is_streamed(hmreq);
 
-		/* Assign the right Vhost for this request. */
-		if (tfw_http_req_set_context(req)) {
-			TFW_INC_STAT_BH(clnt.msgs_otherr);
-			tfw_client_drop(req, 500, "cannot find"
-				      "Vhost for request");
-			return TFW_BLOCK;
-		}
-
-		r = tfw_gfsm_move(&conn->state, TFW_HTTP_FSM_REQ_MSG, &data_up);
-		TFW_DBG3("TFW_HTTP_FSM_REQ_MSG return code %d\n", r);
-		/* Don't accept any following requests from the peer. */
-		if (r == TFW_BLOCK) {
-			TFW_INC_STAT_BH(clnt.msgs_filtout);
-			tfw_client_block(req, 403, "parsed request"
-				       " has been filtered out");
-			return TFW_BLOCK;
-		}
+	/* At least part of request is parsed. */
+	__HTTP_FSM_START(req->state.st) {
 
+	/* A new request is received . */
+	__HTTP_FSM_STATE(Http_Msg_0) {
 		/*
 		 * The time the request was received is used for age
 		 * calculations in cache, and for eviction purposes.
 		 */
 		req->cache_ctl.timestamp = tfw_current_timestamp();
 		req->jrxtstamp = jiffies;
 
+		/* Fall though to Http_Msg_Buffer */
+	}
+
+	/*
+	 * Buffer message before processing it. Headers must be fully parsed
+	 * before processing the message. Even in stream mode.
+	 */
+	__HTTP_FSM_STATE(Http_Msg_Buffer) {
+		int p_stage = tfw_http_parse_stage(hmreq);
+
+		if (unlikely(p_stage == TFW_HTTP_PARSE_HEADERS))
+			__HTTP_FSM_EXIT();
```

Comment:
The stage is called `TFW_HTTP_PARSE_HEADERS`, but since `hm->crlf.flags & TFW_STR_F_COMPLETE` - headers had been parsed already, so: 1. why we need to stay in this state; 2. stage name is somewhat confusing.